airgun/entities/domain.py

```diff
@@ -0,0 +1,192 @@
+from navmazing import NavigateToSibling
+from wait_for import wait_for
+
+from airgun.entities.base import BaseEntity
+from airgun.navigation import NavigateStep, navigator
+from airgun.views.domain import (
+    DomainCreateView,
+    DomainListView,
+    DomainEditView
+)
+
+
+class DomainEntity(BaseEntity):
+    _DEFAULT_ORG = 'Default Organization'
+    _DEFAULT_LOC = 'Default Location'
+    _CREATE_SUCCESSFUL = 'Successfully created {}.'
+    _UPDATE_SUCCESSFUL = 'Successfully updated {}.'
+    _DELETE_SUCCESSFUL = 'Successfully deleted {}.'
+
+    @staticmethod
+    def _submit_domain_form(view, **kwargs):
+        """Create or update domain.
+
+        Form is the same for both cases, we use 1 method here to fill the form.
+        """
+        dns_domain = kwargs.get('dns_domain')
+        full_name = kwargs.get('full_name')
+        dns_capsule = kwargs.get('dns_capsule')
+        parameters = kwargs.get('parameters')
+        locations = kwargs.get('locations')
+        organizations = kwargs.get('organizations')
+
+        # Fill 'Domain' tab
+        if dns_domain:
+            view.domain.dns_domain.fill(dns_domain)
+        if full_name:
+            view.domain.full_name.fill(full_name)
+        if dns_capsule:
+            view.domain.dns_capsule.fill(dns_capsule)
+
+        # Fill 'Parameters' tab
+        if parameters:
+            params = []
+            for param_name, param_value in parameters.items():
+                params.append({'name': param_name, 'value': param_value})
+            view.parameters.params.fill(params)  # fill overwrites old params
+
+        # Fill locations tab
+        if locations:
+            view.locations.multiselect.set_assigned(locations)
+
+        # Fill organizations tab
+        if organizations:
+            view.organizations.multiselect.set_assigned(organizations)
+
+        view.submit_button.click()
+
+    def create(self, dns_domain, **kwargs):
+        """
+        Create a new domain.
+
+        :param dns_domain: str, full DNS domain name
+        :param full_name: (optional) str, full name describing the domain
+        :param dns_capsule: (optional) str of DNS capsule to use
+        :param parameters: (optional) param dict or list of param dicts to set
+            Format: [{'name': str, 'value': str}]
+        :param locations: list of location strings to be assigned.
+            All other locations will be unassigned.
+            Default: ['Default Location']
+        :param organizations: list of org strings to be assigned.
+            All other orgs will be unassigned.
+            Default: ['Default Organization']
+        """
+        view = self.navigate_to(self, 'New')
+
+        kwargs['dns_domain'] = dns_domain
+
+        if not kwargs.get('locations'):
```

Comment:
all that data is also redundant as you should specify nothing in case it was not required on a test level or if such assignment is mandatory then test still should handle that on its own. Yes, that will take more efforts and lines of code for tests, but you will avoid situation when framework do things that it should not do

Reply:
dns_domain is the only required arg, I made it a required arg for creating a domain since it is required in the UI as well. I could make it a non-required kwarg in case people want to test creating a domain without properly filling in dns_domain. The other args cab all be 'None' and nothing in the form will be filled.

I can also change things to not fill locations and organizations with defaults if those are 'None'